### ISSUE TITLE
`@JsonInclude(content=)` not working for `Map`-valued properties via POJO class annotation

### ISSUE THREAD

**Issue by doeiqts** [FasterXML/jackson-databind#1649]
Using the newest release version (2.9.0.pr3, though it also happens in 2.8.8) it seems that using the content() field of @JsonInclude doesn't behave as documented (http://fasterxml.github.io/jackson-annotations/javadoc/2.9.pr3/). Documentation does have conflicting statements though, so it may just be a documentation error.

The main text says that the content() field applies to Lists, Maps, and arrays.

> To base inclusion on value of contained value(s), you will typically also need to specify content() annotation; for example, specifying only value() as JsonInclude.Include.NON_EMPTY for a {link java.util.List} would exclude Lists with no Java elements, but would include Lists with `null` elements. To exclude **Lists** with only nulls, you would use both annotations like so:
>
>public class Bean {
   @JsonInclude(value=Include.NON_EMPTY, content=Include.NON_NULL)
   public **List**<String> entries;
}
>
>Similarly you could further exclude **Lists, Maps or arrays** that only contain "empty" elements, or "non-default" values (see JsonInclude.Include.NON_EMPTY and JsonInclude.Include.NON_DEFAULT for more details).

But the field description only mentions Maps.

>Inclusion rule to use for entries ("content") of annotated **Maps**; defaults to JsonInclude.Include.ALWAYS.

In my testing, it does not seem to actually work with Lists, just Maps. It also only works at the field annotation level, rather than the class annotation level. I would expect that I could specify it in the class annotation and it would apply to everything that it's supposed to in that class.

Examples:

**Class level with tests**
```
@JsonInclude(value=JsonInclude.Include.NON_EMPTY, content=JsonInclude.Include.NON_EMPTY)
public class Bean {
    public String myString;

    public List<String> myList;

    public Map<String, String> myMap;
}
```

```
@Test
    public void test1() throws Exception {
        ObjectMapper objectMapper = new ObjectMapper();
        String expectedResult = "{}";

        String myString = null;

        List<String> myList = null;

        Map<String, String> myMap = null;

        Bean bean = new Bean();
        bean.myString = myString;
        bean.myList = myList;
        bean.myMap = myMap;

        String result = objectMapper.writeValueAsString(bean);

        assertEquals(expectedResult, result);
    }
```
Passes

```
@Test
    public void test2() throws Exception {
        ObjectMapper objectMapper = new ObjectMapper();
        String expectedResult = "{}";

        String myString = null;

        List<String> myList = new ArrayList<>();

        Map<String, String> myMap = new HashMap<>();

        Bean bean = new Bean();
        bean.myString = myString;
        bean.myList = myList;
        bean.myMap = myMap;

        String result = objectMapper.writeValueAsString(bean);

        assertEquals(expectedResult, result);
    }
```
Passes

```
@Test
    public void test3() throws Exception {
        ObjectMapper objectMapper = new ObjectMapper();
        String expectedResult = "{}";

        String myString = null;

        List<String> myList = new ArrayList<>();
        myList.add(null);

        Map<String, String> myMap = new HashMap<>();
        myMap.put("1", null);

        Bean bean = new Bean();
        bean.myString = myString;
        bean.myList = myList;
        bean.myMap = myMap;

        String result = objectMapper.writeValueAsString(bean);

        assertEquals(expectedResult, result);
    }
```
Fails
>Expected :{}
>Actual   :{"myList":[null],"myMap":{"1":null}}

```
@Test
    public void test4() throws Exception {
        ObjectMapper objectMapper = new ObjectMapper();
        String expectedResult = "{}";

        String myString = "";

        List<String> myList = new ArrayList<>();
        myList.add("");

        Map<String, String> myMap = new HashMap<>();
        myMap.put("1", "");

        Bean bean = new Bean();
        bean.myString = myString;
        bean.myList = myList;
        bean.myMap = myMap;

        String result = objectMapper.writeValueAsString(bean);

        assertEquals(expectedResult, result);
    }
```
Fails
>Expected :{}
>Actual   :{"myList":[""],"myMap":{"1":""}}

It works better when changed to a field annotation, but still not for lists. It also seems silly that the class level annotation wouldn't work for content() when it works for value().

**Field level with tests**
```
public class Bean {
    @JsonInclude(value=JsonInclude.Include.NON_EMPTY, content=JsonInclude.Include.NON_EMPTY)
    public String myString;

    @JsonInclude(value=JsonInclude.Include.NON_EMPTY, content=JsonInclude.Include.NON_EMPTY)
    public List<String> myList;

    @JsonInclude(value=JsonInclude.Include.NON_EMPTY, content=JsonInclude.Include.NON_EMPTY)
    public Map<String, String> myMap;
}
```

**Test1**
Passes

**Test2**
Passes

**Test3**
Fails
>Expected :{}
>Actual   :{"myList":[null]}

**Test4**
Fails
>Expected :{}
>Actual   :{"myList":[""]}

So I'm not sure whether the documentation just needs to be cleaned up, or if there's an actual issue with Lists. It also seems like the class level and field level versions should work the same.

**Comment by cowtowncoder** [comment]
Handling of multiple level defaulting set of inclusion criteria is bit difficult to deal with; as there are at least following levels (from lowest to highest precedence)

1. Global default setting (mapper.setDefaultPropertyInclusion())
2. Class annotation (for collection types)
3. Per-type default inclusion (via mapper.configOverride(...))
4. Property defaults (via POJO class annotation (*))
5. Per-property annotation

(*) yes, meaning of class annotation is different for POJO types, container types -- for latter it is for type itself, sort of like per-type override; for POJOs it defines baseline inclusion to use for properties POJO has.

Given this, it is very possible that you may have hit a bug of some kind.
At the same time, combination of rules sometimes leads to complicated resolution, and perhaps precedence might be non-intuitive, but working "as-designed" from Jackson perspective.

So basically I'll have to look into this, once I have bit more time.

Thank you for reporting this!

**Comment by cowtowncoder** [comment]
Hmmh. I should have read description more thoroughly first -- it is actually true that `content` IS NOT considered for `Collection`s or arrays, but only for `Map`s. While it would be good to eventually support it for collections/arrays as well this is not yet implemented.

So I will consider this documentation task at this point -- another issue may be filed for adding actual support for serialized-as-array types (arrays, `Collection`s), for versions post 2.9 (due to timing can't be done for 2.9).

**Comment by doeiqts** [comment]
Fair enough. There was still the issue of field level versus class level for maps though.

**Comment by cowtowncoder** [comment]
@doeiqts Ok thanks. Class-level is... challenging. Its meaning for POJOs is quite specialized, and distinct from how Maps should behave. But I hope to find time to look into that part; at minimum exact behavior should be defined.

**Comment by cowtowncoder** [comment]
Improved Javadocs of `@JsonInclude.content`.

**Comment by cowtowncoder** [comment]
Looks like I can reproduce class annotation issue.

**Comment by cowtowncoder** [comment]
And now I can kind of see how/why. So the problem comes from `PropertyBuilder` -- which has access to surrounding POJO defaults -- only handling `value` (not content), which it is able to handle. But there is no way to pass `content` information, as things are.

Solving the problem is not trivial (i.e not a simple bug/mishandling). I'll see if there is a way to instead try to access information from `createContextual()` of `MapSerializer`: that may be possible.
Fwtw the same problem should affect reference types (`AtomicReference`, `Optional`).

**Comment by mxmlnglt** [comment]
So basically this is the issue here in jackson-databind for this message?

https://github.com/FasterXML/jackson-annotations/commit/82935b6135f78fb70771a4b6a9a7372865abd66f

> Note that `content` is NOT currently (as of Jackson 2.9) supported for arrays or {@link java.util.Collection}s, but supported may be added in future versions.

**Comment by cowtowncoder** [comment]
@mxmlnglt I think so.
